### ISSUE TITLE
Nunit 3.9 Release

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#2540]
I am planning on releasing 3.9 this week. I still need to scan through the open issues and see if there are other high priority issues that should be included. If anyone has candidates, please add them to the milestone and we can discuss here.

**Comment by rprouse** [comment]
@jnm2 you can add your API diff to this issue.

**Comment by jnm2** [comment]
Alright, this is comparing 3.8.1  to 3.9.0-dev-04674:

<details>

```diff
namespace NUnit.Framework
{
    [System.Obsolete("The AssertionHelper class will be removed in a coming release.")]
    public class AssertionHelper
    {
-       public static NUnit.Framework.Constraints.ConstraintExpression Exactly(int expectedCount);
+       public static NUnit.Framework.Constraints.ItemsConstraintExpression Exactly(int expectedCount);

-       public NUnit.Framework.Constraints.EqualConstraint Contains(object expected);
+       public NUnit.Framework.Constraints.SomeItemsConstraint Contains(object expected);

-       public NUnit.Framework.Constraints.RangeConstraint InRange(System.IComparable from, System.IComparable to);
+       public NUnit.Framework.Constraints.RangeConstraint InRange(object from, object to);

-       public NUnit.Framework.Constraints.EqualConstraint Member(object expected);
+       public NUnit.Framework.Constraints.SomeItemsConstraint Member(object expected);
    }

    public class Has
    {
-       public static NUnit.Framework.Constraints.EqualConstraint Member(object expected);
+       public static NUnit.Framework.Constraints.SomeItemsConstraint Member(object expected);
    }

+   [System.AttributeUsage(System.AttributeTargets.Parameter, AllowMultiple = true, Inherited = false)]
-   public class RangeAttribute : ValuesAttribute
+   public class RangeAttribute : DataAttribute, NUnit.Framework.Interfaces.IParameterDataSource
    {
+       public System.Collections.IEnumerable GetData(NUnit.Framework.Interfaces.IParameterInfo parameter);
    }

    public class TestFixtureData : NUnit.Framework.Internal.TestFixtureParameters
    {
+       public TestFixtureData SetName(string name);
    }
}
namespace NUnit.Framework.Constraints
{
    public abstract class CollectionItemsEqualConstraint : CollectionConstraint
    {
+       protected bool IgnoringCase { get; }

+       protected bool UsingExternalComparer { get; }
    }

    public class EqualConstraint : Constraint
    {
+       public EqualConstraint Using<T>(System.Func<T, T, bool> comparer);
    }

    public class SomeItemsConstraint : PrefixConstraint
    {
+       public SomeItemsConstraint Using<TCollectionType, TMemberType>(System.Func<TCollectionType, TMemberType, bool> comparison);

+       public SomeItemsConstraint Using(System.Collections.IComparer comparer);

+       public SomeItemsConstraint Using<T>(System.Collections.Generic.IComparer<T> comparer);

+       public SomeItemsConstraint Using<T>(System.Comparison<T> comparer);

+       public SomeItemsConstraint Using(System.Collections.IEqualityComparer comparer);

+       public SomeItemsConstraint Using<T>(System.Collections.Generic.IEqualityComparer<T> comparer);
    }
}
namespace NUnit.Framework.Interfaces
{
    public class TestOutput
    {
+       public string TestId { get; }

-       public TestOutput(string text, string stream, string testName);
+       public TestOutput(string text, string stream, string testId, string testName);
    }
}
```

</details><br/>

I'm not worried about any one of the five possible breaking changes, but here is an examples for each one for you to eyeball:

<details>

```cs
// CS0029 Cannot implicitly convert type 'NUnit.Framework.Constraints.SomeItemsConstraint' to 'NUnit.Framework.Constraints.EqualConstraint'
EqualConstraint constraint = Has.Member(expected);
```

```cs
// CS0029 Cannot implicitly convert type 'NUnit.Framework.RangeAttribute' to 'NUnit.Framework.ValuesAttribute'
ValuesAttribute attribute = new RangeAttribute(1, 10);
```

```cs
// CS7036 There is no argument given that corresponds to the required formal parameter 'testName' of 'TestOutput.TestOutput(string, string, string, string)'
new NUnit.Framework.Interfaces.TestOutput("text", "stream", "testName");
// (if we have a reason to avoid this one, we could leave the original constructor in place)
```

#### Requires naming obsolete class and getting warning

```cs
public class SomeFixture : AssertionHelper
{
    [Test]
    public void SomeTest()
    {
        // CS0029 Cannot implicitly convert type 'NUnit.Framework.Constraints.ContainsConstraint' to 'NUnit.Framework.Constraints.EqualConstraint'
        EqualConstraint constraint = Contains(null);

        // CS0029 Cannot implicitly convert type 'NUnit.Framework.Constraints.ContainsConstraint' to 'NUnit.Framework.Constraints.EqualConstraint'
        EqualConstraint constraint = Member(null);
    }
}
```

</details>

**Comment by mikkelbu** [comment]
Looks great @jnm2 . I won't worry about the `Has.Member` change prior to https://github.com/nunit/nunit/commit/aa19bd1e5ca27b0b1ee74af03a0d8e0a9f60968e#diff-5bef3e0c3dde4d1fbf1eb895741faab3R198 it returned a `CollectionContainsConstraint` which were removed.

**Comment by rprouse** [comment]
I am planning on getting the release out today. I see one outstanding issue/PR which I am reviewing. If there is anything else that I should be looking at, please ping me directly. I won't be refreshing the milestone to look for new things 😄

**Comment by jnm2** [comment]
There's an outstanding issue I know of that needs to be resolved before 3.9: https://github.com/nunit/nunit/pull/2538#issuecomment-343514937

**Comment by jnm2** [comment]
Diff from [last comment](https://github.com/nunit/nunit/issues/2540#issuecomment-342486434) to 9b2159f (3.9.0-dev-04674):

```diff
namespace NUnit.Framework
{
    public class TestFixtureData : NUnit.Framework.Internal.TestFixtureParameters
    {
+       public TestFixtureData SetName(string name);
    }
}
namespace NUnit.Framework.Constraints
{
    public abstract class CollectionItemsEqualConstraint : CollectionConstraint
    {
+       public CollectionItemsEqualConstraint Using<T>(System.Func<T, T, bool> comparer);
    }

    public class NUnitEqualityComparer
    {
-       public bool AreEqual(object x, object y, ref Tolerance tolerance);
+       public bool AreEqual(object x, object y, ref Tolerance tolerance, bool topLevelComparison = true);
    }

    public class ThrowsExceptionConstraint : Constraint
    {
-       protected override object GetTestObject<TActual>(ActualValueDelegate<TActual> del);
+       public override ConstraintResult ApplyTo<TActual>(ActualValueDelegate<TActual> del);
    }
}
```

Total diff from 3.8.1 to 9b2159f (3.9.0-dev-04674):
<details>

```diff
namespace NUnit.Framework
{
    [System.Obsolete("The AssertionHelper class will be removed in a coming release.")]
    public class AssertionHelper
    {
-       public static NUnit.Framework.Constraints.ConstraintExpression Exactly(int expectedCount);
+       public static NUnit.Framework.Constraints.ItemsConstraintExpression Exactly(int expectedCount);

-       public NUnit.Framework.Constraints.EqualConstraint Contains(object expected);
+       public NUnit.Framework.Constraints.SomeItemsConstraint Contains(object expected);

-       public NUnit.Framework.Constraints.RangeConstraint InRange(System.IComparable from, System.IComparable to);
+       public NUnit.Framework.Constraints.RangeConstraint InRange(object from, object to);

-       public NUnit.Framework.Constraints.EqualConstraint Member(object expected);
+       public NUnit.Framework.Constraints.SomeItemsConstraint Member(object expected);
    }

    public class Has
    {
-       public static NUnit.Framework.Constraints.EqualConstraint Member(object expected);
+       public static NUnit.Framework.Constraints.SomeItemsConstraint Member(object expected);
    }

+   [System.AttributeUsage(System.AttributeTargets.Parameter, AllowMultiple = true, Inherited = false)]
-   public class RangeAttribute : ValuesAttribute
+   public class RangeAttribute : DataAttribute, NUnit.Framework.Interfaces.IParameterDataSource
    {
+       public System.Collections.IEnumerable GetData(NUnit.Framework.Interfaces.IParameterInfo parameter);
    }
}
namespace NUnit.Framework.Constraints
{
    public abstract class CollectionItemsEqualConstraint : CollectionConstraint
    {
+       protected bool IgnoringCase { get; }

+       protected bool UsingExternalComparer { get; }

+       public CollectionItemsEqualConstraint Using<T>(System.Func<T, T, bool> comparer);
    }

    public class EqualConstraint : Constraint
    {
+       public EqualConstraint Using<T>(System.Func<T, T, bool> comparer);
    }

    public class NUnitEqualityComparer
    {
-       public bool AreEqual(object x, object y, ref Tolerance tolerance);
+       public bool AreEqual(object x, object y, ref Tolerance tolerance, bool topLevelComparison = true);
    }

    public class SomeItemsConstraint : PrefixConstraint
    {
+       public SomeItemsConstraint Using<TCollectionType, TMemberType>(System.Func<TCollectionType, TMemberType, bool> comparison);

+       public SomeItemsConstraint Using(System.Collections.IComparer comparer);

+       public SomeItemsConstraint Using<T>(System.Collections.Generic.IComparer<T> comparer);

+       public SomeItemsConstraint Using<T>(System.Comparison<T> comparer);

+       public SomeItemsConstraint Using(System.Collections.IEqualityComparer comparer);

+       public SomeItemsConstraint Using<T>(System.Collections.Generic.IEqualityComparer<T> comparer);
    }

    public class ThrowsExceptionConstraint : Constraint
    {
-       protected override object GetTestObject<TActual>(ActualValueDelegate<TActual> del);
+       public override ConstraintResult ApplyTo<TActual>(ActualValueDelegate<TActual> del);
    }
}
namespace NUnit.Framework.Interfaces
{
    public class TestOutput
    {
+       public string TestId { get; }

-       public TestOutput(string text, string stream, string testName);
+       public TestOutput(string text, string stream, string testId, string testName);
    }
}
```

</details>
